### PR TITLE
Feature/rebuild fully tested

### PR DESCRIPTION
**Pull request by Xorima**
### Description

The current iteration of the cookbook does not expose all the common parts of a rule and has used `-m comment` while match does not prefix with the `-m`. 

This commit totally rewrites the cookbook and is still a work in progress, opening for comments and thoughts

This will also migrate the cookbook to github actions for testing

Things still to do:

- Have resource to handle the install of iptables (default recipe)
- Have a resource to clear the rules (disabled recipe)
- See if we can use the cli instead of a rewrite to disk and reload as a better option (Parsing issues maybe)
- Write some other resources which are wrappers on `iptables_rule`, such as a `iptables_tcp_rule` which will add properties for the options the tcp matcher provides
- Update the readme with all this documentation (What would your thoughts be on having a documentation folder similar to sous-chefs with each resources documented individually?)
### Issues Resolved

[List any existing issues this PR resolves]

### Check List

- [ ] All tests pass. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/TESTING.MD>
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable
- [ ] All commits have been signed for the Developer Certificate of Origin. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/CONTRIBUTING.MD>

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,16 +1,70 @@
...
+        uses: actions/checkout@master
+      - name: Run yaml Lint
+        uses: actionshub/yamllint@master
 
+  mdl:
     runs-on: ubuntu-latest
+    steps:
+      - name: Check out code
+        uses: actions/checkout@master
+      - name: Run Markdown Lint
```
Comment (tas50) [inline .github/workflows/ci.yml]: No markdown or yaml linting for this github org

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,16 +1,70 @@
...
+  dokken:
+    needs: [mdl, yamllint, delivery]
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os:
+          - 'amazonlinux'
+          - 'debian-9'
+          - 'ubuntu-1604'
+          - 'centos-7'
```
Comment (tas50) [inline .github/workflows/ci.yml]: Make sure we don’t lose Centos 8 testing in the move to actions

File: resources/chain.rb [hunk 3]
```diff
@@ -1,47 +1,62 @@
...
+property :cookbook, String,
+          default: 'iptables',
+          description: 'Source cookbook to find the template in'
+
+property :sensitive, [true, false],
+          default: false,
+          description: 'mark the resource as senstive'
+
+property :config_file_full_path, String,
+          default: lazy { default_iptables_rules_file(ip_version) },
```
Comment (tas50) [inline resources/chain.rb]: Can we keep this as config_file for backwards compatibility?

**Comment by Xorima** [comment]
@tas50 will write some docs around this tomorrow if I get time,
What is odd is that while dokken and kitchen pass on my local they do not pass on github actions (Local is fedora) so I am just leaving centos-8 for now, we also have unit tests for this

Documentation plans are:
- Changelog
- Upgrading notes
- Resources documented individually

I have restored as well travis for now (Note, travis also only ran `delivery` so there has never been testing on this cookbook for kitchen/integration testing)

File: resources/chain.rb [hunk 4]
```diff
@@ -1,47 +1,71 @@
...
+          description: 'The full path to find the rules on disk'
 
 action :create do
-  Chef::Resource::Template.send(:include, Iptables::ChainHelpers)
+  # We are using the accumalator pattern here
+  # This is as we are managing a single config file but using multiple
+  # resouces to allow a cleaner api for the end user
+  # Note, this will only ever go as a file on disk at the end of a chef run
+  if new_resource.table.class == 'String'
```
Comment (tas50) [inline resources/chain.rb]: It's probably better to just use a proc to set the string to a symbol and then validate on that.

File: resources/chain.rb [hunk 5]
```diff
@@ -1,47 +1,71 @@
...
+  Chef::Log.warn("Chain #{new_resource.chain} should be a symbol, the property will no longer accept Strings in the next major version") if new_resource.chain.class == 'String'
+
+  table_name = table.to_s
   with_run_context :root do
     edit_resource(:template, new_resource.config_file) do |new_resource|
-      source new_resource.source
+      source new_resource.source_template
       cookbook new_resource.cookbook
       sensitive new_resource.sensitive
-      mode new_resource.filemode
```
Comment (tas50) [inline resources/chain.rb]: any reason to remove this property?

**Comment by Xorima** [comment]
@tas50 ready for a review on this one, will sort out resources for the other recipes if this goes ahead

**Comment by Xorima** [comment]
@tas50 this is all ready for review, note I tried to get `deprecated_property_alias 'target', 'Jump', 'The target property was renamed jump in 7.0.0 and will be removed in 8.0.0'` but it blew up so if you can take a look inside `iptables_rule` I would appreciate it

**Comment by Xorima** [comment]
Notes to self, set `deprecated_property_alias 'target', 'jump', 'The target property was renamed jump in 7.0.0 and will be removed in 8.0.0' ` ... jump should be lower case

Get more testing in this!

File: resources/packages.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
+include Iptables::Cookbook::Helpers
+
+property :packages, Array,
+  default: lazy { package_names },
+  description: 'The packages to install for iptables'
+
+action :installl do
```
Comment (bmhughes) [inline resources/packages.rb]: Typo

**Comment by Xorima** [comment]
Left to fix is Centos-6 as some packages do not exist there we use, and any other tests that break, hopefully I can get this all green tomorrow and ready to merge in with full testing supported.

Also want to extend the unit tests on the helper library for the few additional functions now in there

**Comment by Xorima** [comment]
@bmhughes do you have any idea how to get this working on centos 6, I keep getting modules cannot be loaded errors (note `dokken` does not seem to work, but kitchen we get that far ...)

**Comment by bmhughes** [comment]
@Xorima on your branch I can get CentOS 6 to converge on dokken to the point that it fails with the missing ip6tables service, I haven't got any 6 boxes left but from memory the iptables init file deals with both iptables and ip6tables if the ipv6 modules are loaded so the ip6tables service needs to be skipped.

Where is it failing for you? I don't see any module loaded errors for me, just the service missing exception. I'll try vagrant as well but I expect it'll be the same.

Also I've realised I've leaked my internal gitlab CI in the public dokken config 🤦🏻‍♂️, the volumes can be removed from all the platforms

**Comment by bmhughes** [comment]
@Xorima ignore the last comment I'd help if I was on the right branch 🤦🏻‍♂️.

It converges for me on vagrant, it fails on stopping when trying to unload the iptables kernel modules which I presume is the same error. It failed with module in use so I'm wondering if it's possible to do in docker?

**Comment by Xorima** [comment]
Yeah, the tests also fail mate, just banging my head against a wall, centos 6 support ends in november

**Comment by bmhughes** [comment]
On vagrant it's failing the test because the iptable modules are missing and on docker it won't unload them, it's catch 22. 

The CentOS 6 vagrant image is missing the `ip_conntrack` module if you `modprobe ip_conntrack` then `service iptables restart` the rules will be loaded and the inspec test will then pass. Not sure how best to work around this with, might be easiest to add an execute resource to the test cookbook to load it as it's going to be a bento PR otherwise.

Docker I'm looking into but that seems more difficult with how the networking side of it works.

**Comment by Xorima** [comment]
If you want to take a stab at getting it in you have write access to my repo :)

**Comment by bmhughes** [comment]
> 
> 
> If you want to take a stab at getting it in you have write access to my repo :)

Will do, I have something that may work I want to test with it a bit more tomorrow and I'll push it. Still haven't given up on docker yet.

**Comment by bmhughes** [comment]
That passes for me on Vagrant and Dokken now.

**Comment by bmhughes** [comment]
That needs a serious squash but it passes on CI/dokken/vagrant now.

**Comment by Xorima** [comment]
@bmhughes AMAZING!

Now should that functionality be automatically done by our resources? 

@tas50  enjoy the review ;)

**Comment by bmhughes** [comment]
I don't think so as it's all a workaround for Kitchen on various platforms and I believe it should be relevant for a 'proper' system. Without spinning up a full CentOS 6 VM to test with I never remember having to manually load the modules.

1. Vagrant/Bento the modules aren't loaded (but should be, there's a quirk there with the packer template I guess). I have no idea why the init script doesn't load them if missing, even more helpful that is successfully exits without actually loading the rules.

2. The dokken and azure's images load them correctly but for some reason the CentOS 6 init scripts default to unloading the modules as well as flushing the rules which I don't think will be possible with docker as it's dependant on netfilter for networking afaik so there's various differences there between dokken and a vagrant VM. In a container there's always a dependency on at least one of the modules even with all the rules flushed so they won't unload.

**Comment by tas50** [comment]
I'm ok with retaining the history. It's huge, but this way we don't squash out valuable information later. Thanks for doing all this @Xorima.

**Pull request merged**